index.js

```diff
@@ -36,6 +44,7 @@ function compose (middleware) {
       if (i <= index) return Promise.reject(new Error('next() called multiple times'))
       index = i
       let fn = middleware[i]
+      if (isArray(fn)) fn = compose(fn)
```

Comment:
my concern here is that we compose every time we call dispatch. this might have performance issues.

i'd prefer to flatten the array at `compose()`, but that might not be what you want as you won't be able to edit nested middleware during runtime (which i don't think you should be doing)

Reply:
If I remember correctly, the point was not to edit middleware at runtime, but to allow consumer (Koa) to pass nested/wrapped as an array - and instead of flattening the array I simply opted to compose recursively.

But you're right, I would compose on each dispatch. Maybe minuscule penalty, but definitely unnecessary as pre-flattening would avoid that.

But, this could be easily done by consumer as well. The whole point was to allow Koa to add some clean logic to wrap each middleware, but I'm not convinced this is of concern anymore. If anything, I think this can be closed and forgotten (I already had).